Core/include/Acts/Vertexing/SeedVertexFinder.ipp

```diff
@@ -0,0 +1,656 @@
+      sorted_spacepoints = sortSpacepoints(spacepoints);
+
+  // find triplets
+  std::vector<Acts::SeedVertexFinder<spacepoint_t>::Triplet> triplets =
+      findTriplets(sorted_spacepoints);
+
+  // if no valid triplets found
+  if (triplets.empty()) {
+    return {};
+  }
+
+  Acts::Vector3 vtx = Acts::Vector3::Zero();
+  if (m_cfg.minimalizeWRT == "planes") {
+    // find a point closest to all planes defined by the triplets
+    vtx = findClosestPointFromPlanes(triplets);
+  } else if (m_cfg.minimalizeWRT == "rays") {
+    // find a point closest to all rays fitted through the triplets
+    vtx = findClosestPointFromRays(triplets);
+  } else {
+    ACTS_ERROR("value of minimalizeWRT is "
+               << m_cfg.minimalizeWRT
+               << ", allowed values are \"planes\" or \"rays\" ");
+  }
+
+  return vtx;
+}
+
+template <typename spacepoint_t>
+std::vector<std::vector<
+    std::vector<std::vector<std::pair<spacepoint_t const*, Acts::ActsScalar>>>>>
+Acts::SeedVertexFinder<spacepoint_t>::sortSpacepoints(
+    const std::vector<spacepoint_t>& spacepoints) const {
+  std::vector<std::pair<spacepoint_t const*, Acts::ActsScalar>> helper = {};
+  std::vector<std::vector<std::pair<spacepoint_t const*, Acts::ActsScalar>>>
+      helper_eta(m_cfg.numZSlices, helper);
+
+  std::vector<std::vector<
+      std::vector<std::pair<spacepoint_t const*, Acts::ActsScalar>>>>
+      near_spacepoints(m_cfg.numPhiSlices, helper_eta),
+      middle_spacepoints(m_cfg.numPhiSlices, helper_eta),
+      far_spacepoints(m_cfg.numPhiSlices, helper_eta);
+
+  for (const auto& sp : spacepoints) {
+    // phi will be saved for later
+    Acts::ActsScalar phi = detail::radian_pos(std::atan2(sp.y(), sp.x()));
+    int phislice = (int)(phi / (2 * M_PI) * m_cfg.numPhiSlices);
+
+    if (fabs(sp.z()) >= m_cfg.maxAbsZ) {
+      continue;
+    }
+    int zslice = (int)((sp.z() + m_cfg.maxAbsZ) / (2 * m_cfg.maxAbsZ) *
+                       m_cfg.numZSlices);
+
+    // input spacepoint is sorted into one subset
+    if (sp.r() < m_cfg.rMinMiddle) {
+      if (m_cfg.rMinNear < sp.r() && sp.r() < m_cfg.rMaxNear) {
+        if (std::fmod(m_cfg.useFracPhiSlices * phislice, 1.0) >=
+            m_cfg.useFracPhiSlices) {
+          continue;
+        }
+        near_spacepoints.at(phislice).at(zslice).emplace_back(
+            (spacepoint_t const*)&sp, phi);
+      }
+    } else {
+      if (sp.r() < m_cfg.rMinFar) {
+        if (sp.r() < m_cfg.rMaxMiddle) {
+          if (std::fmod(m_cfg.useFracZSlices * zslice, 1.0) >=
+              m_cfg.useFracZSlices) {
+            continue;
+          }
+          middle_spacepoints.at(phislice).at(zslice).emplace_back(
+              (spacepoint_t const*)&sp, phi);
+        }
+      } else {
+        if (sp.r() < m_cfg.rMaxFar) {
+          far_spacepoints.at(phislice).at(zslice).emplace_back(
+              (spacepoint_t const*)&sp, phi);
+        }
+      }
+    }
```

Comment:
I wonder if that could move more to the left with proper `else if` cases

Reply:
I changed it to `else if` when possible. Also note that logic above leaves "holes", as spacepoints e.g. between `rMaxMiddle` and `rMinFar` are not assigned anywhere.